Fixes #17622: Display need for virt-who in the UI

ehelms opened the pull request:
Uses the virt_limit property of a pool and if its a Red Hat
subscription to determine if a user needs virt-who. This is displayed
on the subscription list, and subscription details. As well, if
a manifest includes subscriptions that require virt-who an info
alert is displayed to point them to the Red Hat virtual instance
guide.

jlsherrill commented on app/models/katello/subscription.rb:
```diff
@@ -33,6 +33,10 @@ def recently_expired?
     end
 
+    def virt_who?
+      pools.any?(&:virt_who)
```
i think this would be a bit more efficient:   
```
pools.where(:virt_who => true).any?
```

jlsherrill commented on app/models/katello/glue/candlepin/owner.rb:
```diff
@@ -21,7 +21,9 @@ def owner_info
       end
 
       def owner_details
```
any thoughts on putting this value outside of owner_details?

jlsherrill replied:
in the rabl i mean

ehelms replied:
I thought about it, and it seemed stranger to put on the organization Rabl itself. The proper solution seemed to have a 'manifest' rabl and API since we talk about it as a first class object but that seemed larger in scope.

jlsherrill replied:
works for me

jlsherrill commented on app/models/katello/glue/candlepin/owner.rb:
```diff
@@ -21,7 +21,9 @@ def owner_info
       def owner_details
-        Resources::Candlepin::Owner.find self.label
+        details = Resources::Candlepin::Owner.find self.label
+        details['virt_who'] = self.subscriptions.any?(&:virt_who?)
```
I'm still a little concerned (maybe less so than below) about the performance of this, depending on the # of subscriptions it could be slow, you could rewrite it like:

```
self.subscriptions.joins(:pools).where("#{Katello::Pool.table_name}.virt_who" => true).any?
```

jlsherrill commented:
This will need an upgrade rake task, correct?

ehelms commented:
I've updated the model level methods to follow your suggestions on being more efficient. Also added some model level tests for the new search method and model methods. I copied the 2.4 upgrade rake task for importing subscriptions to a 3.3 upgrades folder. Assuming this all works as expected, I'll open an installer PR for the upgrade task.

ehelms commented:
@thomasmckay you may want to take a look at this as well

jlsherrill commented on lib/katello/tasks/upgrades/3.3/import_subscriptions.rake:
```diff
@@ -0,0 +1,12 @@
+namespace :katello do
+  namespace :upgrades do
+    namespace '2.4' do
```
3.3

ehelms replied:
Touche - updated

ehelms commented on engines/bastion_katello/app/assets/javascripts/bastion_katello/subscriptions/details/subscription-details.controller.js:
```diff
@@ -51,5 +51,7 @@ angular.module('Bastion.subscriptions').controller('SubscriptionDetailsControlle
             return "";
         };
+
+        $scope.virtWhoToolTip = translate("If the virt-who field is Yes then the subscription requires the use of virt-who. Learn how to configure and use this tool in the <a href=\"https://access.redhat.com/documentation/en/red-hat-satellite/6.2/single/virtual-instances-guide/\">Virtual Instances Guide</a>.");
```
Note this points to the Red Hat Satellite virtual instances guide which is less typical for us to put in, but I don't really know of a better guide than this given its specificity.

jlsherrill replied:
as discussed, i've noticed that this shows up for subscriptions that support virt-who but do not require it (i.e. can be used on a hypervisor to provide guest subs, or on a physical host to provide content).   Thus the statement 'requires virt who' is not correct.  I'm not sure how common this is but may be something look in to.

thomasmckay replied:
I have not been able to get this PR to work, but this would be my suggestion: Change the column title to "virt-who Functionality". In that column, for the primary subscriptions put text like "Yes". For guest subscriptions put "virt-who required".

thomasmckay replied:
@jlsherrill - that's true to a degree but there are two education points: The primary VDC (or other) subscription and what it does, and the guest subscription and how to attach it. I don't _need_ virt-who to use the primary sub, but I am _required_ to run virt-who for guest sub.

ohadlevy replied:
does it make sense to extract the URL to some other variable? allowing others to find it quickly? also consider using something like the documentation helper (e.g. documentation_ur) that exists in foreman today.

ehelms replied:
@ohadlevy we could send the link up in the Bastion config that gets loaded up if you think that would be more robust in the end (https://github.com/Katello/katello/blob/master/engines/bastion_katello/lib/bastion_katello/engine.rb#L37)

bkearney replied:
We chatted with the usability folks about this. The main goal is to point out to users who have just loaded a manifest that they need to install and configure virt-who. This is why it is tied to the host subscription. I would suggest we stay in line with the latest downstream comps which show it as "Virt-Who Required" and then linking to the portal docs. @thomasmckay @ehelms

jlsherrill commented:
Commit message is incorrect (missing #)

ehelms commented:
The commit message was correct - the PR title was just missing the `#`

ehelms commented on lib/katello/engine.rb:
```diff
@@ -278,6 +278,7 @@ def find_assets(args = {})
       load "#{Katello::Engine.root}/lib/katello/tasks/upgrades/3.0/delete_docker_v1_content.rake"
       load "#{Katello::Engine.root}/lib/katello/tasks/upgrades/3.0/update_puppet_repository_distributors.rake"
       load "#{Katello::Engine.root}/lib/katello/tasks/upgrades/3.0/update_subscription_facet_backend_data.rake"
+      load "#{Katello::Engine.root}/lib/katello/tasks/upgrades/3.3/import_subscriptions.rake"
```
@jlsherrill I noticed this was missing and updated the PR with it.

ehelms commented on app/models/katello/glue/candlepin/pool.rb:
```diff
@@ -123,6 +123,8 @@ def import_data
           pool_attributes[:unmapped_guest] = true
         end
 
+        pool_attributes[:virt_who] = pool_json['virt_limit'] != "0" && subscription.redhat?
```
Updated

ehelms commented:
@jlsherrill @thomasmckay Where have we landed with this?

ohadlevy commented on app/models/katello/pool.rb:
```diff
@@ -26,6 +26,7 @@ class Pool < Katello::Model
     scoped_search :on => :consumed, :complete_value => true
     scoped_search :on => :account_number, :complete_value => true, :rename => :account
     scoped_search :on => :contract_number, :complete_value => true, :rename => :contract
+    scoped_search :on => :virt_who, :complete_value => true
```
should this be only explicit search? (for search performance)
I guess the same is true for other search attributes.

ehelms replied:
Is explicit search a scoped search option? If yes, sounds good to me.

ohadlevy replied:
its: ```:only_explicit => true```

ohadlevy commented on app/models/katello/subscription.rb:
```diff
@@ -17,6 +17,10 @@ def self.with_subscribable_content
     end
 
+    def self.virt_who
+      joins(:pools).where("#{Katello::Pool.table_name}.virt_who" => true)
```
should this be a scope?

ehelms replied:
Technically this is a scope (http://guides.rubyonrails.org/active_record_querying.html#scopes) I can covert this to use the `scope` method if that would make it more clear.

ohadlevy commented on app/models/katello/subscription.rb:
```diff
@@ -33,6 +37,14 @@ def recently_expired?
       pools.any?(&:recently_expired?)
     end
 
+    def virt_who
```
i find the class and instance methods on the same name confusing, is it on purpose ?

jlsherrill replied:
i'd probably propose something like 'using_virt_who' for the class method, and 'virt_who_pools' for the subscription method (or something similar)

ohadlevy commented on db/migrate/20161209162947_add_virt_who_to_katello_pools.rb:
```diff
@@ -0,0 +1,9 @@
+class AddVirtWhoToKatelloPools < ActiveRecord::Migration
+  def up
+    add_column :katello_pools, :virt_who, :boolean, :default => false, :null => false
```
should we also add an index?

ehelms replied:
Fair question, I'm always hesitant to add them out of the gate to reduce index explosion. However, we tend to be a read heavy and not a write heavy application so I can add one if we think this would gain us performance.

ehelms commented:
@jlsherrill @thomasmckay please re-review

ehelms commented:
[test]

thomasmckay commented on app/models/katello/glue/candlepin/pool.rb:
```diff
@@ -126,6 +126,8 @@ def import_data
           pool_attributes[:unmapped_guest] = true
         end
 
+        pool_attributes[:virt_who] = pool_json['virt_limit'] != "0" && !pool_json['virt_limit'].nil? && subscription.redhat?
```
```
pool_attributes[:virt_who] = pool_attributes[:virt_limit] != "0" && !pool_attributes[:virt_limit].nil? && subscription.redhat?
```

thomasmckay replied:
:baby: :baby_bottle: :shipit:

ehelms commented:
@jlsherrill @thomasmckay updated to use `pool_attributes` to pull the virt_limit

thomasmckay commented:
@ehelms You're going to want to use symbol rather than string
```
pool_attributes[:virt_who] = pool_attributes[:virt_limit] != "0" && !pool_attributes[:virt_limit].nil? && subscription.redhat?
  ```

thomasmckay commented:
APJ

jlsherrill commented:
APJ from me too
